### PR TITLE
Sparse Implementation of `build_coleman_forest` plus a fix for bottleneck import

### LINKED ISSUE #310: Bottleneck should not raise a warning unless relevant function in treeple.stats is called
These lines seem like they are ran regardless of whether or not treeple.stats is used, since they are in the global namespace.

```
if BOTTLENECK_AVAILABLE and DISABLE_BN_ENV_VAR not in os.environ:
    nanmean_f = bn.nanmean
    anynan_f = lambda arr: bn.anynan(arr, axis=2)
else:
    warnings.warn(
        "Not using bottleneck for calculations involvings nans. Expect slower performance."
    )
    nanmean_f = np.nanmean
    anynan_f = lambda arr: np.isnan(arr).any(axis=2)
```

so I get this warning consistently no matter what

```
/Users/adam2392/Documents/treeple/treeple/stats/utils.py:34: UserWarning: Not using bottleneck for calculations involvings nans. Expect slower performance.
  warnings.warn(
```

@ryanhausen do you mind submitting a PR to patch this behavior?

### PR DESCRIPTION
**Pull request by ryanhausen**
#### Reference Issues/PRs

Fixes #310 

#### What does this implement/fix? Explain your changes.

This adds a sparse implementation of `build_coleman_forest` to `treeple.stats`. The sparse implementation relies on `scipy.stats.sparse` and so only works in the case of binary classification and regression. In my tests using `cProfile`/`memray`, the sparse implementation is a little over 50% faster and uses about %7 less memory, see [pdf](https://github.com/user-attachments/files/16681054/sparse_profiling.pdf) for more information. I am also attaching the [profiling data](https://github.com/user-attachments/files/16681063/sparse_profiling.zip)
 if you'd like to take a look.

This PR also includres a change to the bottleneck implementation to for the dense implementation of `build_coleman_forest`. Specifically there are two important changes:
1. The check for the existence of the bottleneck package is now done using `importlib.util.find_spec("bottleneck")`. The old check using `sys` seemed to work fine and passed tests, but actually didn't seem work unless bottleneck had been imported before, which when the tests are run is true, but may not be the case in other situations
2. the aliased and lambda functions `nanmean_f` and `anynan_f` are now defined using `def` so that the warning message can moved into their respective calls rather than at import.

#### Any other comments?

The scipy.stats sparse implementation should be swapped out for pydata.sparse when that implementation is performant enough, as it is more general.

### CODE REVIEW

File: treeple/stats/forest.py
Comment (adam2392) [inline treeple/stats/forest.py]: Will the returned predictions and posteriors also be "sparse"? If so, perhaps we can document this. I assume structurally they are the same and can be converted via `todense` and `to_csc`?

File: treeple/stats/utils.py
Comment (adam2392) [inline treeple/stats/utils.py:47]: It seems this is not tested? Perhaps we can use the environment variable setting to ensure the warning gets triggered when `bottleneck` is not used?

File: treeple/stats/utils.py
Comment (adam2392) [inline treeple/stats/utils.py]: Can you document the input and output in a docstring?

File: treeple/stats/utils.py
Comment (adam2392) [inline treeple/stats/utils.py]:
```suggestion
    # XXX: Because we are assuming 2D sparse matrices for binary classification or
```

File: treeple/stats/utils.py
Comment (adam2392) [inline treeple/stats/utils.py]: Do we first construct `coo` and then convert to `csc` for any particular reason? Perhaps good to state it in an in-line comment.

**Review by adam2392 — COMMENTED** [review]
Cool! Thanks for the profiling data and documentation. It's very convincing. 

Left a few comments to tighten up the PR

File: treeple/stats/forest.py
Comment (ryanhausen) [reply treeple/stats/forest.py]: Ah good catch this should be documented. Currently they are sparse csc matrices, but actually that behavior will be different than the dense behavior because the dense matrix has nans and the sparse doesn't. I'll change it to return dense matrices that are the same as the dense implementation.

File: treeple/stats/utils.py
Comment (ryanhausen) [reply treeple/stats/utils.py]: good catch, the csc matrix can be instantiated using `coo` format, so I'll simplify those constructors

File: treeple/stats/utils.py
Comment (ryanhausen) [reply treeple/stats/utils.py:47]: This is weird. For some reason when I run this through pytest this code doesn't seem to be acknowledged, even `print` statements don't get run. If I run the function manually e.g. something like:

```
if __name__ == "__main__":
    test_build_coleman_forest(False, False)
```

then the behavior seems to be correct. I am going to work on it. I may just split the test out. I'll follow up when I have a fix.

Comment (ryanhausen) [reply treeple/stats/utils.py:47]: Ok they should be covered now. I tested it separately.

**Comment by ryanhausen** [comment]
@adam2392 I have updated the PR, let me know if I missed anything or if you want to see some more changes.

File: treeple/stats/forest.py
Comment (adam2392) [inline treeple/stats/forest.py:188]: Can you explain what's going on here? I guess it is to make the return in "dense" format?

File: treeple/stats/tests/test_forest.py
Comment (adam2392) [inline treeple/stats/tests/test_forest.py:256]:
```suggestion
    # to allow testing without bottleneck
    importlib.reload(utils)
    importlib.reload(stats)
```

Is this accurate?

**Review by adam2392 — COMMENTED** [review]
Mostly lgtm. Just some small nits

File: treeple/stats/tests/test_forest.py
Comment (ryanhausen) [reply treeple/stats/tests/test_forest.py:256]: yep, I added a comment

File: treeple/stats/forest.py
Comment (ryanhausen) [reply treeple/stats/forest.py:188]: Yes, I added a comment. Your understanding is correct. Let me know if you would like more clarification.

**Comment by ryanhausen** [comment]
@adam2392 I added some documentation to address your comments. Let me know what you think.

**Review by adam2392 — APPROVED** [review]

**Comment by adam2392** [comment]
I will let @sampan501 @SUKI-O @YuxinB @PSSF23 take a look as well.

File: treeple/stats/utils.py
Comment (sampan501) [inline treeple/stats/utils.py:334]: maybe I missed it somewhere, but are there unit tests for these methods anywhere?

Comment (adam2392) [reply treeple/stats/utils.py:334]: @ryanhausen

Comment (ryanhausen) [reply treeple/stats/utils.py:334]: @sampan501, @adam2392  No there isn't a test for this function in test_utils.py. This function, `_get_forest_preds_sparse`, is a good candidate for a unit test though. I can write one up. 

I worry the other functions: `_parallel_build_null_forests_sparse` and `_compute_null_distribution_coleman_sparse`, like their dense counterparts, do too much to be easily unit tested and are tested via the integration tests. However,  if you would like, we can refactor the dense/sparse functions into smaller units so that they can all have tests written. Would probably be cleaner, but would take time. 

What would you like to see?

Comment (adam2392) [reply treeple/stats/utils.py:334]: For now, I think it's fine the way it is. If we find means of re-designing the internal API to make things simpler and easier to test, then it's warranted, but I don't see a clear path. Do you?

Comment (ryanhausen) [reply treeple/stats/utils.py:334]: I think it would take some thought. There is some repetition we might be able to leverage, but I am not sure if it would come at the cost of clarity. So just write up a test for `_get_forest_preds_sparse`?

Comment (adam2392) [reply treeple/stats/utils.py:334]: Yeah I think that's sufficient then for now.

Comment (sampan501) [reply treeple/stats/utils.py:334]: sounds good. Maybe we can make an issue for having such unit tests/re-designing the API so we don't forget. Otherwise, I think this PR lgtm

**Comment by ryanhausen** [comment]
@sampan501  @adam2392  Hey I added a test, but it looks like I some other tests are failing. I submitted an issue, #328. I'll start looking into it, but it might go faster if you have an intuition on what the issue is.

**Comment by adam2392** [comment]
This is gonna be an issue due to parameter validation in scikit-learn. I will push a fix when I have time. It seems the rest of the code works tho.

I would prioritize other issues until I have the time to push up the fix

**Comment by ryanhausen** [comment]
@adam2392 sounds good. Should we let this sit and update and merge after you have a fix for the other issue or merge as is? @sampan501

**Comment by sampan501** [comment]
I'm happy to merge if @adam2392 is cool with it

**Review by sampan501 — APPROVED** [review]

**Comment by adam2392** [comment]
> @adam2392 sounds good. Should we let this sit and update and merge after you have a fix for the other issue or merge as is? @sampan501

I will merge once the fix is added and the CI is confirmed to be happy. No rush on this one.

**Comment by adam2392** [comment]
@ryanhausen if you update wrt `main`, the CIs should be fixed

**Comment by adam2392** [comment]
Thanks @ryanhausen !

**Pull request merged**
